inmoose/limma/contrasts.py

```diff
@@ -0,0 +1,262 @@
+# -----------------------------------------------------------------------------
+# Copyright (C) 2024 Maximilien Colange
+
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+# -----------------------------------------------------------------------------
+
+import copy
+import logging
+import numpy as np
+import patsy
+import pandas as pd
+import scipy
+
+from ..utils import Factor, cov2cor
+
+
+def makeContrasts(contrasts, levels):
+    """
+    Construct the contrast matrix corresponding to specified contrasts of a set of parameters.
+
+    This function expresses contrasts between a set of parameters as a numeric
+    matrix. The parameters are usually the coefficients from a linear model
+    fit, so the matrix specifies which comparisons between the coefficients are
+    to be extracted from the fit. The output from this function is usually used
+    as input to :func:`contrasts_fit`.
+
+    Arguments
+    ---------
+    contrasts : list
+        strings parseable as expressions that describe the contrasts
+    levels : list
```

Comment:
In the following code, levels can also be a `patsy.DesignMatrix`, and contrast not a `list`

Reply:
fixed